### PR TITLE
Feature: Network Assistance and Dynamic Policies

### PR DESCRIPTION
**Pull request by davidjwbbc**
# Overview

These changes introduce the Network Assistance and the Dynamic Polices features to the AF as well as including a uplift to comply with the latest published specification version, TS 26.512 V17.6.0.

# Network Assistance

The Network Assistance feature has been added but will contains the delivery boost option, the bit rate recommendation (also known as throughput estimation) is still under development.

## Activating

The Network Assistance feature is disabled by default in the 5GMS Application Function. It is enabled by setting both the `msaf.open5gsIntegration` and `msaf.offerNetworkAssistance` settings in the `msaf.yaml` configuration file to `true`.

## Delivery boost

When invoked at reference point M5, the reference implementation applies a fixed guaranteed minimum downlink bit rate for a fixed period of time.

- The guaranteed bit rate for the boost defaults to "1 Mbps". This can be overridden by setting the `msaf.networkAssistance.deliveryBoost.minDlBitRate` parameter in the `msaf.yaml` configuration file.
- The boost period defaults to 30 seconds. This can be changed by setting a different number of seconds in the `msaf.networkAssistance.deliveryBoost.boostPeriod` parameter in the `msaf.yaml` configuration file.

## Bit rate recommendation

This Network Assistance feature is not yet released. Due to lack of support in the Open5GS PCF implementation, we haven't been able to fully implement and test it.

# Dynamic Policies

Dynamic Policies will be advertised in the Service Access Information when Policy Templates are configured for the provisioning session using the API at reference point M1.

In this implementation Policy Templates immediately transition into the __Ready__ state as soon as they are defined. This behaviour will remain until a suitable API is introduced to allow network operators to query and approve/deny policy templates.

# Other changes
- The `Configuration` class has been moved out of the m1-session tool and into the rt_m1_client module so that it can be easily shared between the python applications (currently used by the m1-session and msaf-configuration tools).
- The Open5GS OpenAPI templates have been replaced for the AF specific object models with versions which will honour read-only attributes and min/max values for integer ranges and will also return a reason for serialisation errors.
  - This introduces function calls for convertRequestToJSON(), convertResponseToJSON(), parseRequestFromJSON() and parseResponseFromJSON() to process or ignore properties marked as read-only in the relevant situations.
  - All 5GMS AF specific models are now prefixed with `msaf_api_` to distinguish them from the Open5GS OpenAPI models.

### CODE REVIEW

File: src/5gmsaf/dynamic-policy.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
+/*
+License: 5G-MAG Public License (v1.0)
+Author: Dev Audsin
+Copyright: (C) 2022 British Broadcasting Corporation
```
Comment (rjb1000) [inline src/5gmsaf/dynamic-policy.h:4]: 2023

File: src/5gmsaf/local.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
+/*
+License: 5G-MAG Public License (v1.0)
+Copyright: (C) 2022 British Broadcasting Corporation
```
Comment (rjb1000) [inline src/5gmsaf/local.h:3]: 2023

File: src/5gmsaf/network-assistance-delivery-boost.c [hunk 3]
```diff
@@ -0,0 +1,45 @@
+/*
+License: 5G-MAG Public License (v1.0)
+Copyright: (C) 2022 British Broadcasting Corporation
```
Comment (rjb1000) [inline src/5gmsaf/network-assistance-delivery-boost.c:3]: 2023

File: src/5gmsaf/network-assistance-delivery-boost.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
+/*
+License: 5G-MAG Public License (v1.0)
+Copyright: (C) 2022 British Broadcasting Corporation
```
Comment (rjb1000) [inline src/5gmsaf/network-assistance-delivery-boost.h:3]: 2023

File: src/5gmsaf/network-assistance-session.h [hunk 5]
```diff
@@ -0,0 +1,79 @@
+/*
+License: 5G-MAG Public License (v1.0)
+Author: Dev Audsin
+Copyright: (C) 2022 British Broadcasting Corporation
```
Comment (rjb1000) [inline src/5gmsaf/network-assistance-session.h:4]: 2023

File: src/5gmsaf/openapi-templates/model-header.mustache
Comment (rjb1000) [inline src/5gmsaf/openapi-templates/model-header.mustache:1]: Shouldn't this file have a copyright statement added?

File: src/5gmsaf/pcf-session.c [hunk 6]
```diff
@@ -0,0 +1,44 @@
+/*
+License: 5G-MAG Public License (v1.0)
+Copyright: (C) 2022 British Broadcasting Corporation
```
Comment (rjb1000) [inline src/5gmsaf/pcf-session.c:3]: 2023

File: src/5gmsaf/pcf-session.c
Comment (rjb1000) [inline src/5gmsaf/pcf-session.c:1]: Missing author name.

File: src/5gmsaf/pcf-session.h [hunk 7]
```diff
@@ -0,0 +1,34 @@
+/*
+License: 5G-MAG Public License (v1.0)
+Author: Dev Audsin
+Copyright: (C) 2022 British Broadcasting Corporation
```
Comment (rjb1000) [inline src/5gmsaf/pcf-session.h:4]: 2023

File: src/5gmsaf/policy-template.h [hunk 8]
```diff
@@ -0,0 +1,53 @@
+/*
+License: 5G-MAG Public License (v1.0)
+Copyright: (C) 2022 British Broadcasting Corporation
```
Comment (rjb1000) [inline src/5gmsaf/policy-template.h:3]: 2023

File: src/5gmsaf/timer.c
Comment (rjb1000) [inline src/5gmsaf/timer.c:1]: Why isn't this file included by reference from the Open5GS project?

File: src/5gmsaf/timer.h [hunk 9]
```diff
@@ -0,0 +1,35 @@
+/*
+License: 5G-MAG Public License (v1.0)
+Copyright: (C) 2022 British Broadcasting Corporation
```
Comment (rjb1000) [inline src/5gmsaf/timer.h:3]: 2023

File: tests/msaf/pcf-cache-test.h
Comment (rjb1000) [inline tests/msaf/pcf-cache-test.h:1]: Missing author name.

**Review by rjb1000 — CHANGES_REQUESTED** [review]
Found a few missing author names, and a few cases where the copyright year on new files was 2022 instead of 2023.

In addition, any files started in 2022 but touched in 2023 should have "2022-2023" as their copyright year.

File: src/5gmsaf/network-assistance-delivery-boost.h (hunk 4 above)
Comment (rjb1000) [reply src/5gmsaf/network-assistance-delivery-boost.h:3]: Also missing author name.

File: src/5gmsaf/network-assistance-delivery-boost.c (hunk 3 above)
Comment (rjb1000) [reply src/5gmsaf/network-assistance-delivery-boost.c:3]: Also missing author name.

File: src/5gmsaf/local.h (hunk 2 above)
Comment (rjb1000) [reply src/5gmsaf/local.h:3]: Also missing author name.

File: src/5gmsaf/policy-template.h (hunk 8 above)
Comment (rjb1000) [reply src/5gmsaf/policy-template.h:3]: Also missing author name.

File: src/5gmsaf/timer.h (hunk 9 above)
Comment (rjb1000) [reply src/5gmsaf/timer.h:3]: Also missing author name.

File: src/5gmsaf/timer.c
Comment (davidjwbbc) [reply src/5gmsaf/timer.c:1]: Each NF has it's own timer.c file. It's there to handle our local 5GMS AF timers.

File: tests/msaf/pcf-cache-test.h
Comment (davidjwbbc) [reply tests/msaf/pcf-cache-test.h:1]: done

File: src/5gmsaf/timer.h (hunk 9 above)
Comment (davidjwbbc) [reply src/5gmsaf/timer.h]: done

File: src/5gmsaf/timer.c
Comment (davidjwbbc) [reply src/5gmsaf/timer.c:1]: I've modified the header to show that it originates from Open5GS but with changes we've added to handle 5GMS AF timer events.

File: src/5gmsaf/policy-template.h (hunk 8 above)
Comment (davidjwbbc) [reply src/5gmsaf/policy-template.h]: done

File: src/5gmsaf/pcf-session.h (hunk 7 above)
Comment (davidjwbbc) [reply src/5gmsaf/pcf-session.h]: done

**Review by rjb1000 — APPROVED** [review]
Thanks for the changes, @davidjwbbc.

**Pull request merged**